linq4j/src/main/java/org/apache/calcite/linq4j/tree/Primitive.java

```diff
@@ -412,69 +410,75 @@ static BigDecimal checkOverflow(BigDecimal value, int precision, int scale) {
-      @Nullable Long value, int precision, int scale, BigDecimal unitScale) {
+      @Nullable Long value, int precision, int scale,
+      BigDecimal unitScale, RoundingMode roundingMode) {
     if (value == null) {
       return null;
     }
     // Divide with the scale expected of the result
-    BigDecimal result = new BigDecimal(value).divide(unitScale, scale, RoundingMode.DOWN);
-    return checkOverflow(result, precision, scale);
+    BigDecimal result = new BigDecimal(value).divide(unitScale, scale, roundingMode);
+    return checkOverflow(result, precision, scale, roundingMode);
   }
 
   /**
    * Convert a long time interval to a decimal value.
    * Called from BuiltInMethod.LONG_INTERVAL_DECIMAL_CAST.
    * @param unitScale Scale describing source interval type */
   public static @Nullable Object longIntervalToDecimalCast(
-      @Nullable Integer value, int precision, int scale, BigDecimal unitScale) {
+      @Nullable Integer value, int precision, int scale,
+      BigDecimal unitScale, RoundingMode roundingMode) {
     if (value == null) {
       return null;
     }
     // Divide with the scale expected of the result
-    BigDecimal result = new BigDecimal(value).divide(unitScale, scale, RoundingMode.DOWN);
-    return checkOverflow(result, precision, scale);
+    BigDecimal result = new BigDecimal(value).divide(unitScale, scale, roundingMode);
+    return checkOverflow(result, precision, scale, roundingMode);
   }
 
   /** Called from BuiltInMethod.DECIMAL_DECIMAL_CAST */
   public static @Nullable Object decimalDecimalCast(
-      @Nullable BigDecimal value, int precision, int scale) {
+      @Nullable BigDecimal value, int precision, int scale, RoundingMode roundingMode) {
     if (value == null) {
       return null;
     }
-    return checkOverflow(value, precision, scale);
+    return checkOverflow(value, precision, scale, roundingMode);
   }
 
   /** Called from BuiltInMethod.INTEGER_DECIMAL_CAST */
   public static @Nullable Object integerDecimalCast(
-      @Nullable Number value, int precision, int scale) {
+      @Nullable Number value, int precision, int scale, RoundingMode roundingMode) {
     if (value == null) {
       return null;
     }
     final BigDecimal decimal = new BigDecimal(value.longValue());
-    return checkOverflow(decimal, precision, scale);
+    return checkOverflow(decimal, precision, scale, roundingMode);
   }
 
   /** Called from BuiltInMethod.FP_DECIMAL_CAST */
   public static @Nullable Object fpDecimalCast(
-      @Nullable Number value, int precision, int scale) {
+      @Nullable Number value, int precision, int scale, RoundingMode roundingMode) {
     if (value == null) {
       return null;
     }
     final BigDecimal decimal = BigDecimal.valueOf(value.doubleValue());
-    return checkOverflow(decimal, precision, scale);
+    return checkOverflow(decimal, precision, scale, roundingMode);
+  }
+
+  public @Nullable Object numberValue(Number value) {
```

Comment:
This looks like a dangerous API. 
If you insist on having this function, I would call it numberValueRoundDown to make it clear what it does.